server/src/main/java/org/opensearch/cluster/coordination/JoinHelper.java

```diff
@@ -352,6 +359,13 @@ public void handleException(TransportException exp) {
                         FailedJoinAttempt attempt = new FailedJoinAttempt(destination, joinRequest, exp);
                         attempt.logNow();
                         lastFailedJoinAttempt.set(attempt);
+                        if (exp instanceof RemoteTransportException && (exp.getCause() instanceof NodeDecommissionedException)) {
```

Comment:
The current flow is:
1. Node sends the JoinRequest to active leader
2. Active leader sends the ValidateJoinRequest
3. Node checks for decommissioning and will fail the request if node has decommissioning attribute

The above logic will not work if the current master eligible node is not active leader and is a candidate accumulating joins  [Code Ref](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/cluster/coordination/Coordinator.java#L603)
```
 if (stateForJoinValidation.nodes().isLocalNodeElectedClusterManager()) {
    .....
   sendValidateJoinRequest(stateForJoinValidation, joinRequest, joinCallback);
 } else {
    processJoinRequest(joinRequest, joinCallback);
   }
```

Suggestion:
Ideally the decommissioning check should be executed at
1. During handleJoinRequest in the Coordinator. This is lightweight check DiscoveryNode would be sending attribute information during join. This will reject any joins from decommissioned node early on itself.
2. During JoinExecutor submittask as well as that will ensure correctness. It will never allow a node with decommission attribute to join the cluster.

Reply:
In that case, can you please help me understand that when current leader eligible node is not active leader and is just accumulating joins, we just skip the validator code flow?

Reply:
> During handleJoinRequest in the Coordinator. This is lightweight check DiscoveryNode would be sending attribute information during join. This will reject any joins from decommissioned node early on itself.

This makes sense. Will update

Reply:
Made these changes. 
1. A check in Coordinator#handleJoinRequest (when the state doesn't have `STATE_NOT_RECOVERED_BLOCK`)
2. Check in the execute method of JoinTask
3. Removed it from builtin validators

@shwetathareja , lmk if this looks ok

Reply:
Hey, I explored more on join validators - 
 
> During handleJoinRequest in the Coordinator. This is lightweight check DiscoveryNode would be sending attribute information during join. This will reject any joins from decommissioned node early on itself.

I think we can keep it as built in join validators, and Coordinator#handleJoinRequest validates all the checks here. So handle join request would be covered here. Another case I see where keeping it at BuiltinValidator would help is during handlePublishRequest. Ideally we would never hit this case as it runs only on Mode.Leader but I think for any such cases in future, it would be good to have all the validators at one place

Reply:
Although I agree, we have to keep it at execute method of the task as it would be the ultimate source of truth. Hence I think putting it in built in validator and execute method should be good

Reply:
Built validator delays the check which can be checked early on during handleJoinRequest itself. Can you help me understand the handlePublishRequest scenario?

Reply:
```
transportService.connectToNode(joinRequest.getSourceNode(), ActionListener.wrap(ignore -> {
            final ClusterState stateForJoinValidation = getStateForClusterManagerService();

            if (stateForJoinValidation.nodes().isLocalNodeElectedClusterManager()) {
                onJoinValidators.forEach(a -> a.accept(joinRequest.getSourceNode(), stateForJoinValidation));
                if (stateForJoinValidation.getBlocks().hasGlobalBlock(STATE_NOT_RECOVERED_BLOCK) == false) {
                    // we do this in a couple of places including the cluster update thread. This one here is really just best effort
                    // to ensure we fail as fast as possible.
                    JoinTaskExecutor.ensureMajorVersionBarrier(
                        joinRequest.getSourceNode().getVersion(),
                        stateForJoinValidation.getNodes().getMinNodeVersion()
                    );
                    // we are checking source node commission status here to reject any join request coming from a decommissioned node
                    // even before executing the join task to fail fast
                    JoinTaskExecutor.ensureNodeCommissioned(joinRequest.getSourceNode(), stateForJoinValidation.metadata());
                }
                sendValidateJoinRequest(stateForJoinValidation, joinRequest, joinCallback);
            } else {
                processJoinRequest(joinRequest, joinCallback);
            }
        }, joinCallback::onFailure));
```

This is the snippet from handleJoinRequest of Coordinator. If you see, this method is also running the same built in validators `onJoinValidators.forEach(a -> a.accept(joinRequest.getSourceNode(), stateForJoinValidation));`. And hence the concern with early checking is actually already resolved here

Reply:
We might never actually hit the `handlePublishRequest` scenario as this would be run only for `Leader` mode. But what I was trying to say is plugins put their join validators in `onJoinValidators` and then Coordinator adds all other validators to this `onJoinValidators`. And then in handlePublishRequest and handleJoinRequest we validate with whatever validators  `onJoinValidators` has. Hence, keeping all the validators centrally in `onJoinValidators` might be good

Reply:
handlePublishRequest is executed on the node joining the master. But with the check inside ```handleJoinRequest``` master will reject the joins upfront.

Reply:
Correct. I meant the built in validators are also executed with handleJoinRequest